### ISSUE TITLE
Decorators where DOM content length != actual content length (question)

### ISSUE THREAD

**Issue by aviraldg** [facebook/draft-js#515]
Hello. I was trying to create a decorator to swap out unicode emoji for images like :smile:

I tried a basic (naive) implementation using a decorator but it fails terribly when trying to modify content, probably because the length of the DOM content does not match the actual content (which consists of several codepoints)

Am I missing an obvious way to do this?

**Comment by iandoe** [comment]
Use or have a look at how it's done here : https://www.draft-js-plugins.com/plugin/emoji

**Comment by aviraldg** [comment]
So for emoji I did think of a similar approach (I suspect it works because the content length is actually the same, just that this content is hidden), but that breaks down for user mentions:

![screenshot from 2016-06-15 00-32-45](https://cloud.githubusercontent.com/assets/238916/16642529/8f4ebcd8-4429-11e6-94ae-4add9796f3d8.png)

**Comment by tleunen** [comment]
Have you solve your initial issue with the emoji? I'm currently trying the same thing, but unlike draft-js-plugins Emoji, I'd like to use a svg instead of a background image. But as you said, it seems Draft is all messed up when the dom is more than just `props.children` :/

**Comment by aviraldg** [comment]
@tleunen You _can_ use an SVG as a background image.

**Comment by tleunen** [comment]
Yep but my svg is actually a spritesheet so I believe I have to use the custom block

**Comment by villesau** [comment]
Something what you could do is to add element with contentEditable={false}. This way the internals of that dom is not part of content editable. It might still cause problems with cursor positioning in some cases though.

**Comment by lan9** [comment]
We come across the same issue, our approach is similar to @iandoe suggested: using a regex to identify start/end index of emojis and convert them to immutable entities. It works with SVG background.
